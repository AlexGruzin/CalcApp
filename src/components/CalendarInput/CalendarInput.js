import React, { useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import DatePicker from './DatePicker';

import './CalendarInput.styles.less';

const CalendarInput = ({ label, name, isOpen: isOpenDefault, onChange, start }) => {
  const [isOpen, setIsOpen] = useState(isOpenDefault);

  const handleOnChange = data => {
    onChange(data);
  };

  const toggleOpen = () => {
    // event.stopPropagation();
    // event.preventDefault(); // TODO check node to close via icon
    setIsOpen(!isOpen);
  };

  return (
    <div className="calendar-input">
      {label && (
        <label className="calendar-input__label" htmlFor={name}>
          {label}
        </label>
      )}
      <div className={classNames('calendar-input__inner', isOpen, { withLabel: label })}>
        {/* // TODO wrap with button */}
        {/* eslint-disable-next-line jsx-a11y/no-noninteractive-element-to-interactive-role */}
        <img src="/svg/calendar.svg" alt="calendar-svg" role="button" onClick={toggleOpen} />
        <DatePicker
          handleDateChange={handleOnChange}
          withCalendarIcon={false}
          onSelect={toggleOpen}
          open={isOpen}
          onClickOutside={toggleOpen}
          start={start}
        />
      </div>
    </div>
  );
};

CalendarInput.propTypes = {
  name: PropTypes.string,
  start: PropTypes.instanceOf(Date),
  label: PropTypes.string,
  isOpen: PropTypes.bool,
  onChange: PropTypes.func,
};

CalendarInput.defaultProps = {
  label: '',
  name: '',
  isOpen: false,
  onChange: () => undefined,
  start: new Date(),
};

export default CalendarInput;
