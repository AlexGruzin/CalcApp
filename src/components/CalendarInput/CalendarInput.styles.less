.calendar-input .withLabel.calendar-input__inner {
  margin-top: var(--space-x3);
}

.calendar-input label {
  font-family: "MontserratBold", sans-serif;
  letter-spacing: 0.25px;
  font-size: var(--fontSize-tiny);
}

.calendar-input {
  display: flex;
  flex-direction: column;
  position: relative;
}

.calendar-input__inner {
  display: flex;
  align-items: center;
}

.calendar-input__inner > img {
  position: absolute;
  right: var(--space-x2);
  z-index: 2;
  padding: 7px;
  cursor: pointer;
}

.calendar-input__text {
  display: flex;
  max-width: 200px;
  flex-direction: column;
}

.calendar-input .custom-date-picker__input {
  padding-left: var(--space-x7);
}
