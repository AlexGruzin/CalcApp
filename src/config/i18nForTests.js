import i18n from 'i18next';

// locales
import US from '../locales/en';

i18n.init({
  debug: false,
  fallbackLng: ['US'],
  lng: 'US',

  ns: ['auth', 'resetPassword'],
  defaultNS: 'auth',

  resources: {
    US,
  },
});

export default i18n;

