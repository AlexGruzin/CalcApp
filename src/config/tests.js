import React from 'react';
import { render } from '@testing-library/react';
import { I18nextProvider } from 'react-i18next'
import i18n from "./i18nForTests";
import jsdom from 'jsdom';
const { JSDOM } = jsdom;

/*
 * jest-dom
 * https://github.com/testing-library/jest-dom
 */
import '@testing-library/jest-dom/extend-expect';

/*
 * jest-fetch-mock
 * https://github.com/jefflau/jest-fetch-mock
 */
require('jest-fetch-mock').enableMocks();

if (global.document) {
  global.document.createRange = () => ({
    setStart: () => {
    },
    setEnd: () => {
    },
    commonAncestorContainer: {
      nodeName: 'BODY',
      ownerDocument: document,
    },
    createContextualFragment: str => JSDOM.fragment(str),
    // createContextualFragment: str => document.createElement('hr'),
  });
  global.open = () => undefined;
}

export const renderWithi18next = Component => {
  const Comp = React.cloneElement(Component, {
    changeLanguage: lng => {
      i18n.changeLanguage(lng);
      rerender(<I18nextProvider i18n={i18n}>{Comp}</I18nextProvider>);
    }
  });
  const defaultRender = render(
    <I18nextProvider i18n={i18n}>{Component}</I18nextProvider>
  );
  const { rerender } = defaultRender;
  return defaultRender;
};


// /**
//  * Wraps a component with an Intl Provider.
//  * @param {Object} component The component to render
//  */
// const withTestIntl = component => <IntlProvider locale="en">{component}</IntlProvider>;
//
// /*
//  * https://testing-library.com/docs/example-react-intl
//  */
// export const renderWithReactIntl = component => render(withTestIntl(component));

