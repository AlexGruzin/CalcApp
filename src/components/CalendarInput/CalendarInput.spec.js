import React from 'react';
import { fireEvent, waitFor, render } from '@testing-library/react';
import { renderWithi18next } from 'config/tests';

import CalendarInput from './CalendarInput';

describe('CalendarInput tests', () => {
  it('Check snapshots of CalendarInput are equal.', () => {
    const { asFragment } = render(<CalendarInput start={new Date('1995-12-17T03:24:00')} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('Menu could be opened by mouse click.', async () => {
    const { getByRole, container } = render(<CalendarInput />);

    const icon = getByRole('button');
    fireEvent.click(icon);

    await waitFor(() => {
      expect(container.querySelector('.react-datepicker-popper')).toBeInTheDocument();
    });
  });

  it('New date could be selected from opened menu.', async () => {
    const mockOnChange = jest.fn();
    const { getByText, container } = renderWithi18next(<CalendarInput isOpen onChange={mockOnChange} />);

    await waitFor(() => {
      expect(container.querySelector('.react-datepicker-popper')).toBeInTheDocument();
    });

    const option = getByText('15'); // 15 - day number of some month
    fireEvent.click(option);

    expect(mockOnChange).toHaveBeenCalled();
  });
});
