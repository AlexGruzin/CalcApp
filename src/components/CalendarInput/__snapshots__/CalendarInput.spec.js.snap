// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarInput tests Check snapshots of CalendarInput are equal. 1`] = `
<DocumentFragment>
  <div
    class="calendar-input"
  >
    <div
      class="calendar-input__inner"
    >
      <img
        alt="calendar-svg"
        role="button"
        src="/svg/calendar.svg"
      />
      <div
        class="date-picker__wrapper"
      >
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class="custom-date-picker__input"
              name=""
              placeholder=""
              type="text"
              value="17/12/1995"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
